### PR TITLE
Robust Intersect and Contains

### PR DESCRIPTION
**Pull request by rmanoka**
Use robust predicates in Intersects and Contains traits

+ use robust preds (doesn't need Float anymore)
+ define interior, boundary, and exterior of geometry types (similar to geos/jts)
+ define Intersect and Contains using DE-9IM model
+ fix contains logic in a few corner cases

Future work:

The polygon contains line / polygon logic is not yet fully correct.   I've added a few TODO comments, and an ignored test for cases where our predicate doesn't match the DE-9IM definition.  Unfortunately, it doesn't seem to be a simple change to fix these corner cases.  For instance, geos seems to use a concept called `GeometryGraph` for these cases which I'm yet to understand.

### CODE REVIEW

**Comment by rmanoka** [comment]
A couple of clarifications needed:

1. These two tests seem inconsistent with definition of `Contains` which says it does not include the boundary.
    https://github.com/georust/geo/blob/a6a3218823c440e9f6d52406b9630fea5cf7f395/geo/src/algorithm/contains.rs#L778
    https://github.com/georust/geo/blob/a6a3218823c440e9f6d52406b9630fea5cf7f395/geo/src/algorithm/contains.rs#L785

2.  Implementation of `Rect: Contains<Point>` also seems to allow the point to be on the edges.  Again, not consistent with definition, and implementation on `Polygon`.  

Also checked against shapely/geos which seems to concur with our definition (so points on edges / vertices of the polygon are not considered to be "contained").  Shall we go with the definition (so fix `Rect` and `Triangle`)?

@michaelkirk @urschrei

**Comment by michaelkirk** [comment]
> Shall we go with the definition (so fix Rect and Triangle)?

Yes. Apparently I wrote that broken triangle implementation 😞: b3944caaa

For my own benefit, I referred to http://portal.opengeospatial.org/files/?artifact_id=25355 (thanks @urschrei :wink:)

> **Contains**
>   `a.Contains(b) ⇔ b.Within(a)`

So it's defined it terms of `within`, which itself is defined as:

> **Within**
> The Within relationship is defined as
> `a.Within(b) ⇔ (a∩b=a) ∧ (I(a)∩E(b)=∅)`
> Expressed in terms of the DE-9IM:
>   ```
> a.Within(b) ⇔ [ I(a)∩I(b)≠∅ ∧ I(a)∩E(b)=∅ ∧ B(a)∩E(b)=∅ ]
>       ⇔ a.Relate(b, “T*F**F***”) 
> ```

And from the [jts impl](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L858):
```
contains
...
[T*****FF*]
```

So for `a.contains(b)`, boundary contact is not sufficient; The *interior* of `b` must reside (wholly) within the *interior* of `a`.

In the future, I'll be careful to be rigorous about the cases represented in DE-9IM.

**Comment by rmanoka** [comment]
Yeah, DE-9IM based rigorous definitions of our geometry types, and traits definitely seems useful. I believe jts, geos (hence postgis), shapely all use the same definition of Intersects, and Contains so makes sense for us to with the same for interop with the community.  Will add some doc to the types trying to formalize the definitions.

File: geo-types/src/line_string.rs [hunk 1]
```diff
@@ -2,9 +2,25 @@ use crate::{Coordinate, CoordinateType, Line, Point, Triangle};
...
+/// boundary.
+///
+/// # Validity
+///
+/// A `LineString` is valid if it is either empty or
+/// contains 2 or more coordinates. Further, a closed
+/// `LineString` must not self intersect. Note that the
+/// validity is not enforced, and the operations and
+/// predicates are undefined on invalid linestrings.
+///
```
Comment (michaelkirk) [inline geo-types/src/line_string.rs:23]: Thanks for continuing to fleshing out the docs. :+1:

File: geo/src/algorithm/intersects/mod.rs [hunk 2]
```diff
@@ -1,278 +1,114 @@
...
-use num_traits::Float;
-
-/// Checks if the geometry A intersects the geometry B.
-
+use crate::*;
+
+/// Checks if the geometry A intersects the geometry B. More
+/// formally, either boundary or interior of A has non-empty
+/// (set-theoretic) intersection with the boundary or
+/// interior of B. In other words, the [DE-9IM] intersection
```
Comment (michaelkirk) [inline geo/src/algorithm/intersects/mod.rs:6]: Thanks for documenting this! I hope we can get all the traits to this level of specification.

File: geo/src/algorithm/intersects/mod.rs [hunk 3]
```diff
@@ -1,278 +1,114 @@
...
+mod polygon;
+mod rect;
+
+// Since `Intersects` is symmetric, we use a macro to
+// implement `T: Intersects<S>` if `S: Intersects<T>` is
+// available. As a convention, we provide explicit impl.
+// whenever the Rhs is a "simpler geometry" than the target
+// type, and use the macro for the reverse impl.
+#[macro_use]
+macro_rules! symmetric_intersects_impl {
```
Comment (michaelkirk) [inline geo/src/algorithm/intersects/mod.rs:52]: Nice!

File: geo/src/algorithm/intersects/mod.rs [hunk 4]
```diff
@@ -1,278 +1,114 @@
...
 
-impl<T> Intersects<Line<T>> for Polygon<T>
-where
-    T: Float,
-{
-    fn intersects(&self, line: &Line<T>) -> bool {
-        line.intersects(self)
-    }
-}
+symmetric_intersects_impl!(Line<T>, LineString<T>, HasKernel);
```
Comment (michaelkirk) [inline geo/src/algorithm/intersects/mod.rs:72]: Not a big deal, so take it or leave it, but for someone trying to find how the intersection logic is actually implemented, it might be helpful to collocate the macro invocation next to the implementation.

i.e. this one could go next to the `impl<T> Intersects<Line<T>> for LineString<T>`

Another benefit to organizing it that way is it might be easier for the implementor (or reviewer 😉)to audit that each `impl` is preceded by the macro invocation than it is to cross reference that the sum of all appropriate calls exists in this file.

File: geo/src/utils.rs [hunk 5]
```diff
@@ -83,43 +84,83 @@ pub enum CoordPos {
...
+            coord,
+            Coordinate {
+                x: max_x,
+                y: coord.y,
+            },
+        );
+        if ray.intersects(&line) {
+            crossings += 1;
         }
+        eprintln!("current crossing @ iter {} = {}", i, crossings);
```
Comment (michaelkirk) [inline geo/src/utils.rs:163]: did you mean to commit the `eprintln!`?

File: geo/src/algorithm/intersects/polygon.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        if self.exterior().intersects(linestring)
+            || self
+                .interiors()
+                .iter()
+                .any(|inner| inner.intersects(linestring))
+        {
+            true
+        } else {
+            // or if it's contained in the polygon
+            linestring.points_iter().any(|point| self.contains(&point))
```
Comment (michaelkirk) [inline geo/src/algorithm/intersects/polygon.rs:56]: Can we iterate over the coords to avoid building a Point in the point iter? Not a big deal, so take it or leave it.

File: geo/src/algorithm/intersects/polygon.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    fn intersects(&self, p: &Point<T>) -> bool {
+        self.intersects(&p.0)
+    }
+}
+
+impl<T> Intersects<Line<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn intersects(&self, p: &Line<T>) -> bool {
```
Comment (michaelkirk) [inline geo/src/algorithm/intersects/polygon.rs:33]: nit: can we call this `l` (or `line`) instead of `p` since it's a **L**ine?

**Review by michaelkirk — APPROVED** [review]
I have yet to review the `Contains` portion, but I ran out of time for now. 

I have reviewed the `Intersecs` and other changes, and they look great! Just some optional feedback for clarity. 

I'll finish reviewing up hopefully yet tonight, or if not, tomorrow.

File: geo-types/src/line.rs [hunk 8]
```diff
@@ -1,6 +1,9 @@
 use crate::{Coordinate, CoordinateType, Point};
 
-/// A line segment made up of exactly two [`Point`s](struct.Point.html).
+/// A line segment made up of exactly two
+/// [`Point`s](struct.Point.html). The interior and
```
Comment (frewsxcv) [inline geo-types/src/line.rs:4]:
```suggestion
/// [`Coordinate`s](struct.Coordinate.html). The interior and
```

File: geo-types/src/line_string.rs (hunk 1 above)
Comment (frewsxcv) [reply geo-types/src/line_string.rs:23]: +1

File: geo/src/algorithm/contains/mod.rs [hunk 9]
```diff
@@ -1,434 +1,48 @@
...
-        self.contains(&point.0)
-    }
-}
+mod geometry;
+mod line;
+mod line_string;
+mod point;
+mod polygon;
+mod rect;
+mod triangle;
```
Comment (frewsxcv) [inline geo/src/algorithm/contains/mod.rs:45]: good idea splitting this file up into multiple modules 🏆

File: geo/src/algorithm/intersects/mod.rs (hunk 4 above)
Comment (rmanoka) [reply geo/src/algorithm/intersects/mod.rs:72]: Good point; will do.

File: geo/src/utils.rs (hunk 5 above)
Comment (rmanoka) [reply geo/src/utils.rs:163]: Oops!

**Comment by rmanoka** [comment]
Thanks for the comments @frewsxcv @michaelkirk ; I have addressed them, also added a few lines in `CHANGES.md` to highlight the recent robust pred. changes.  Will wait a couple more days to see there are any minor fixes we spot, and merge this PR.

closes #415

File: geo-types/CHANGES.md [hunk 10]
```diff
@@ -1,5 +1,10 @@
 # Changes
 
+## Master (not released)
```
Comment (michaelkirk) [inline geo-types/CHANGES.md:3]: Yes, thank you for doing this! Tracking as we go will make the eventual release easier.

File: geo/src/algorithm/contains/polygon.rs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+where
+    T: HasKernel,
+{
+    fn contains(&self, p: &Point<T>) -> bool {
+        self.contains(&p.0)
+    }
+}
+
+// TODO: ensure DE-9IM compliance: esp., when
+// line.start and line.end is on the boundaries
```
Comment (michaelkirk) [inline geo/src/algorithm/contains/polygon.rs:34]: Right as your TODO alludes, it seems like the current logic would fail for a case like:

image

Where the line touches, but does not pass, the interior.
Or if the line touches, but does not pass, the exterior.

File: geo/src/algorithm/contains/polygon.rs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        // both endpoints are contained in the polygon and the line
+        // does NOT intersect the exterior or any of the interior boundaries
+        self.contains(&line.start)
+            && self.contains(&line.end)
+            && !self.exterior().intersects(line)
+            && !self.interiors().iter().any(|inner| inner.intersects(line))
+    }
+}
+
+// TODO: also check interiors
```
Comment (michaelkirk) [inline geo/src/algorithm/contains/polygon.rs:49]: Right, this one also looks wrong. I wouldn't want to include new broken implementations, but I believe you are just moving these, right?

File: geo/src/algorithm/contains/polygon.rs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+// ┌──────────────────────────────────┐
+// │ Implementations for MultiPolygon │
+// └──────────────────────────────────┘
+impl<G, T> Contains<G> for MultiPolygon<T>
+where
+    T: CoordinateType,
+    Polygon<T>: Contains<G>,
```
Comment (michaelkirk) [inline geo/src/algorithm/contains/polygon.rs:86]: It looks like we're newly accepting *any* type which implements the Contains trait, whereas before MultiPolygon only supported Coordinate and Point, is that right?

Does this existing implementation work for types with more dimensionality than a point?

e.g. If the two squares are part of a single multipolygon, even though the black line is contained by neither square individually, is it contained by the multipolygon?

image

File: geo/src/algorithm/contains/rect.rs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    }
+}
+
+impl<T> Contains<Rect<T>> for Rect<T>
+where
+    T: CoordinateType,
+{
+    fn contains(&self, bounding_rect: &Rect<T>) -> bool {
+        // All points of LineString must be in the polygon ?
+        self.min().x <= bounding_rect.min().x
```
Comment (michaelkirk) [inline geo/src/algorithm/contains/rect.rs:35]: So a degenerate Rect which is either a Line or a Point will be `contained` if it's on the boundary - I don't think that's correct, is it?

File: geo/src/algorithm/contains/triangle.rs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+// └──────────────────────────────┘
+
+impl<T> Contains<Coordinate<T>> for Triangle<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, coord: &Coordinate<T>) -> bool {
+        let ls = LineString(vec![self.0, self.1, self.2, self.0]);
+        use utils::*;
+        coord_pos_relative_to_ring(*coord, &ls) == CoordPos::Inside
```
Comment (michaelkirk) [inline geo/src/algorithm/contains/triangle.rs:16]: Thank you for fixing this!

File: geo/src/algorithm/contains/line.rs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+// │ Implementations for Line │
+// └──────────────────────────┘
+
+impl<T> Contains<Coordinate<T>> for Line<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, coord: &Coordinate<T>) -> bool {
+        if self.start == self.end {
+            &self.start == coord
```
[the remark this answers is not in the document]
Comment (rmanoka) [reply geo/src/algorithm/contains/line.rs:16]: Yeah, it is confusing, but that is the precise expected behavior.  I cross verified with shapely (geos), and it does the same.

File: geo/src/algorithm/contains/polygon.rs (hunk 11 above)
Comment (rmanoka) [reply geo/src/algorithm/contains/polygon.rs:34]: Yeah.  Also see the test `line_in_polygon_edgecases` in `contains/mod.rs`.  We might be able to use the winding order trait and a few additions to contains to solve these cases, but wanted to attempt this in a different PR.

File: geo/src/algorithm/contains/polygon.rs (hunk 12 above)
Comment (rmanoka) [reply geo/src/algorithm/contains/polygon.rs:49]: Yeah, these implementations were just moved into this file.  We should fix the two traits for de9im compliance in another pr

File: geo/src/algorithm/contains/polygon.rs (hunk 13 above)
Comment (rmanoka) [reply geo/src/algorithm/contains/polygon.rs:86]: So, [MultiPolygon](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/MultiPolygon.html) should not contain overlapping constituent polys: 

>As per the OGC SFS specification, the Polygons in a MultiPolygon may not overlap, and may only touch at single points. This allows the topological point-set semantics to be well-defined.

With the above defn., I think we can lift any Polygon: Contains<T> to work on MultiPolygon

File: geo/src/algorithm/contains/rect.rs (hunk 14 above)
Comment (rmanoka) [reply geo/src/algorithm/contains/rect.rs:35]: Good catch!  The previous implementation was the same with strict checking, which would also miss some edge cases.  Want to address these in a different PR (needs a few extra impls, utils).   Let me add a todo comment here so we get back to it.

**Comment by michaelkirk** [comment]
Ok, this LGTM. 

I agree it’s fine to address the pre-existing issues in a follow up PR.

Thanks for making this better! I’d say merge this whenever you’re ready.

**Comment by michaelkirk** [comment]
Ah, I didn’t realize polygons in a multi-polygon couldn’t overlap. Thank you for explaining.

**Comment by rmanoka** [comment]
> Ah, I didn’t realize polygons in a multi-polygon couldn’t overlap. Thank you for explaining.

Actually, you're right.  The polygons can touch at a single point which introduces some edge cases.  For eg. two rectangles touching at a single point, and a line passing through that point.  It won't be in any of the individual polygons but is contained in the multipolygon.

Will leave a comment, and we can get to this when we ensure correctness of these two traits.

**Comment by rmanoka** [comment]
bors r=michaelkirk
